framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C

```diff
@@ -56,6 +63,84 @@ MultiAppGeneralFieldNearestNodeTransfer::MultiAppGeneralFieldNearestNodeTransfer
                "are (currently) built with data from multiple applications.");
 }
 
+void
+MultiAppGeneralFieldNearestNodeTransfer::initialSetup()
+{
+  MultiAppGeneralFieldTransfer::initialSetup();
+
+  // Handle the source types ahead of time
+  auto source_types = getParam<std::vector<MooseEnum>>("source_type");
+  _source_is_nodes.resize(_from_var_names.size());
+  _use_zero_dof_for_value.resize(_from_var_names.size());
+  if (source_types.size() != _from_var_names.size())
+    mooseError("Not enough source types specified for this number of variables. Source types must "
+               "be specified for transfers with multiple variables");
+  for (auto var_index : make_range(_from_var_names.size()))
+  {
+    // Get some info on the source variable
+    FEProblemBase & from_problem = *_from_problems[0];
+    MooseVariableFieldBase & from_var =
+        from_problem.getVariable(0,
+                                 _from_var_names[var_index],
+                                 Moose::VarKindType::VAR_ANY,
+                                 Moose::VarFieldType::VAR_FIELD_ANY);
+    System & from_sys = from_var.sys().system();
+    const auto & fe_type = from_sys.variable_type(from_var.number());
+
+    // Select where to get the origin values
+    if (source_types[var_index] == "nodes")
+      _source_is_nodes[var_index] = true;
+    else if (source_types[var_index] == "centroids")
+      _source_is_nodes[var_index] = false;
+    else
+    {
+      // "Nodal" variables are continuous for sure at nodes
+      if (from_var.isNodal())
+        _source_is_nodes[var_index] = true;
+      // for everyone else, we know they are continuous at centroids
+      else
+        _source_is_nodes[var_index] = false;
+    }
+
+    // Some variables can be sampled directly at their 0 dofs
+    if ((_source_is_nodes[var_index] && fe_type.family == LAGRANGE) ||
+        (!_source_is_nodes[var_index] && fe_type.family == MONOMIAL && fe_type.family == CONSTANT))
```

Comment:
@roystgnr 

I m guessing I m too restrictive here. But there's also 20 something element types and 10 families 
I ll see if I can find a good output to see this

Reply:
ok I think I just need to take a deep dive here
https://mooseframework.inl.gov/docs/doxygen/libmesh/generic__projector_8h_source.html

Reply:
Even `LAGRANGE` is tricky - if you have a superparametric element (1st order FE on 2nd order Elem), which is super popular (e.g. the pressure variable in Taylor-Hood), you can't just evaluate component 0 at non-vertex nodes.

Any C0-continuous (including C1-continuous) scalar element in libMesh will take a value of nodal-DoF-0 on vertices.  Any discontinuous `CONSTANT` element will take a value of elemental-DoF-0 everywhere.  Isoparametric `LAGRANGE` will take a value of nodal-DoF-0 on non-vertex nodes.  Vector-valued elements do something similar component-wise but with offsets for different components.  And I think that may be an exhaustive list.

Reply:
oh that first item is going to be tricky

Reply:
ok was not too bad.

So I m correct in assuming that at all orders, the value at the centroid of a MONOMIAL is the same as the dof 0 value (in my mind, all the shape functions except the constant one are 0 there)

Reply:
No, I'm afraid.  On elements with linear mappings, this should be true, but even an element as simple as a trapezoid will be skewed differently.  I think with a bilinear quad you'll see the dof 0 value at the `vertex_average()` point, but not at the `true_centroid()` point in general.

Reply:
Oh, wait, even the above isn't pessimistic enough!

Reply:
So, `MONOMIAL` elements will take their DoF-0 value at `Point(0,0,0)` in master space.  For Edge/Quad/Hex elements, our master elements are just tensor products of the interval `[-1,1]`, and `(0,0,0)` is the centroid of that, and if they have linear (affine) mappings to physical space then the centroid of the mapped element is the mapping of the centroid, and it all works out.  If the mapping isn't linear, then the vertex average isn't the centroid, as I said.

But if the master element isn't a tensor product element?  We don't even use the same convention!  A triangle, for example, has vertices `(0,0)`, `(1,0)`, and `(0,1)` in master space.  Not only does DoF 0 not end up evaluating on the centroid, it barely ends up evaluating on the element at all!  The physical element's vertex 0 ends up being where DoF 0's value matches the physical value.

Reply:
Prisms see master (0,0,0) in the middle of one edge, Pyramids see it in the middle of the square face ... I think the safest thing is to just do the weighted shape function sum for `p>0`; by the time you've tested `has_affine_map()` and `type()` of an `Elem` to see if that's safe to shortcut, you've already done half as much work anyway.

Reply:
removed monomials from the allow-list